File: typescript/packages/core/src/core/gmail/read.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import { mountKey, mountPrefixOf } from '../../utils/key_prefix.ts'
import type { GmailAccessor } from '../../accessor/gmail.ts'
import type { IndexCacheStore } from '../../cache/index/store.ts'
import { entryOrWarm } from '../../cache/index/warm.ts'
import { PathSpec } from '../../types.ts'
import { getAttachment, getMessageRaw, messageJsonBytes } from './messages.ts'
import { readdir } from './readdir.ts'
import { gnuDirname } from '../../utils/path.ts'
import { eisdir, enoent } from '../../utils/errors.ts'

export async function read(
  accessor: GmailAccessor,
  path: PathSpec,
  index?: IndexCacheStore,
): Promise<Uint8Array> {
  const prefix = mountPrefixOf(path.virtual, path.resourcePath)
  const key = path.resourcePath
  if (index === undefined) throw enoent(path.virtual)
  const virtualKey = prefix !== '' ? `${prefix}/${key}` : `/${key}`
  const parentKey = gnuDirname(virtualKey)
  const entry = await entryOrWarm(
    index,
    virtualKey,
    parentKey !== virtualKey
      ? () => readdir(accessor, PathSpec.fromStrPath(parentKey, mountKey(parentKey, prefix)), index)
      : null,
  )
  if (entry === null) throw enoent(path.virtual)
  const rt = entry.resourceType
  if (rt === 'gmail/label' || rt === 'gmail/date' || rt === 'gmail/attachment_dir') {
    throw eisdir(path.virtual)
  }
  if (rt === 'gmail/attachment') {
    const parentResult = await index.get(parentKey)
    if (parentResult.entry === undefined || parentResult.entry === null) {
      throw enoent(path.virtual)
    }
    return getAttachment(accessor.tokenManager, parentResult.entry.id, entry.id)
  }
  const raw = await getMessageRaw(accessor.tokenManager, entry.id)
  return messageJsonBytes(raw)
}

File: python/mirage/core/gsheets/read.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import json
import posixpath

from functools import partial
from mirage.cache.index.warm import entry_or_warm
from mirage.accessor.gsheets import GSheetsAccessor
from mirage.cache.index import NULL_INDEX, IndexCacheStore
from mirage.core.gsheets._client import TokenManager, google_get, sheets_base
from mirage.core.gsheets.readdir import readdir
from mirage.types import PathSpec
from mirage.utils.errors import enoent
from mirage.utils.key_prefix import mount_key, mount_prefix_of

GRID_DATA_PARAM = "true"


async def read_spreadsheet(token_manager: TokenManager,
                           spreadsheet_id: str) -> bytes:
    """Fetch full spreadsheet JSON, cell values included.

    `spreadsheets.get` returns no grid data unless asked, so without
    `includeGridData` the rendered `.gsheet.json` is tab metadata and
    nothing an agent can read a cell from.

    Args:
        token_manager (TokenManager): manages OAuth2 tokens.
        spreadsheet_id (str): Google Sheets spreadsheet ID.

    Returns:
        bytes: JSON response as bytes.
    """
    url = f"{sheets_base(token_manager)}/spreadsheets/{spreadsheet_id}"
    data = await google_get(token_manager, url,
                            {"includeGridData": GRID_DATA_PARAM})
    return json.dumps(data, ensure_ascii=False, separators=(",", ":")).encode()


async def read_values(token_manager: TokenManager, spreadsheet_id: str,
                      range_: str) -> bytes:
    """Read cell values via Values API. Returns JSON array.

    Args:
        token_manager (TokenManager): manages OAuth2 tokens.
        spreadsheet_id (str): Google Sheets spreadsheet ID.
        range_ (str): A1 notation range.

    Returns:
        bytes: JSON response as bytes.
    """
    base = sheets_base(token_manager)
    url = f"{base}/spreadsheets/{spreadsheet_id}/values/{range_}"
    data = await google_get(token_manager, url)
    return json.dumps(data, ensure_ascii=False, separators=(",", ":")).encode()


async def read(
    accessor: GSheetsAccessor,
    path: PathSpec,
    index: IndexCacheStore = NULL_INDEX,
) -> bytes:
    virtual = path.virtual
    prefix = mount_prefix_of(path.virtual, path.resource_path)
    key = path.resource_path
    virtual_key = prefix + "/" + key if prefix else "/" + key
    parent_key = posixpath.dirname(virtual_key) or "/"
    parent_path = PathSpec.from_str_path(parent_key,
                                         mount_key(parent_key, prefix))
    warm = (partial(readdir, accessor, parent_path, index)
            if parent_key != virtual_key else None)
    entry = await entry_or_warm(index, virtual_key, warm)
    if entry is None:
        raise enoent(virtual)
    return await read_spreadsheet(accessor.token_manager, entry.id)
